Add NoSchemaIntrospectionCustomRule

danielrearden opened the pull request:
#2597

IvanGoncharov commented on src/validation/rules/NoIntrospectionFieldsRule.js:
```diff
@@ -0,0 +1,32 @@
+): ASTVisitor {
+  return {
+    Field(node: FieldNode) {
+      if (node.name.value === '__schema' || node.name.value === '__type') {
```
Please check that you are in root type, `context` has functions for that.

danielrearden replied:
@IvanGoncharov Fixed!

IvanGoncharov commented:
@danielrearden Thanks, looks very tidy and match the style of other rules 👍 

I give it more though and technically `__typename` is also introspection field so we can't call it `NoIntrospectionFields` 😞 
Also, we have other fields that can leak similar info for example Apollo federation has `_service` that returns SDL.
So I more incline towards more generic rules like: 
* https://eslint.org/docs/rules/no-restricted-properties
* https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-types.md
* etc.

Also, I think in this case being able to customize error messages would be very helpful.

But I'm not sure if I'm overcomplicating it or not so I need more time to think.
Meanwhile, if you have time you can work on #2601 which is very similar to this one.

IvanGoncharov commented:
@danielrearden Actually scratch it this rule is very short so config for the more generic rule would be more longer that rule itself.
So back the initial question, `__typename` is also introspection field so we need to change rule name.
How about `BanIntrospectionQueriesRule`?

IvanGoncharov commented on src/validation/rules/NoIntrospectionFieldsRule.js:
```diff
@@ -0,0 +1,35 @@
+    Field(node: FieldNode) {
+      if (
+        context.getSchema().getQueryType() === context.getParentType() &&
+        (node.name.value === '__schema' || node.name.value === '__type')
```
It's better to use `getFieldDef` and compare it against constants from `src/type/instospection`.

IvanGoncharov replied:
that way you will have single check

danielrearden replied:
> How about `BanIntrospectionQueriesRule`?

That makes sense. Would `NoIntrospectionQueriesRule` be more inline with the naming of the existing rules?

danielrearden replied:
So a name like `NoFieldsWithIntrospectionTypesRule` would maybe be more accurate.

IvanGoncharov replied:
Technically you right. But it's better to use names that improve discoverability/readability than a name 100% correct. 
For example, `SDL` is an incorrect abbreviation since specification defines `Type System Definition Language` but we use `SDL` anyway since it what community settled on.

I think the spirit of this rule is that we forbidding "introspection queries" so it's better to reflect it a name even through `{ __typename }` is also introspection query.

IvanGoncharov commented on src/validation/rules/NoIntrospectionFieldsRule.js:
```diff
@@ -0,0 +1,35 @@
+// @flow strict
```
better to move it to `rules/extra` directory.

IvanGoncharov replied:
or if you can come up with a better name for this directory.

danielrearden commented:
> How about `BanIntrospectionQueriesRule`?

@IvanGoncharov  That makes sense. Would `NoIntrospectionQueriesRule` be more inline with the existing rule names?

IvanGoncharov commented on src/validation/__tests__/NoIntrospectionFieldsRule-test.js:
```diff
@@ -0,0 +1,100 @@
+import { expectValidationErrors } from './harness';
+
+function expectErrors(queryStr) {
+  return expectValidationErrors(NoIntrospectionFieldsRule, queryStr);
```
it's better to use `expectValidationRulesWithSchema` and have a short SDL snippet in this file than to depend on a huge schema.
I was planning to break "harness" into multiple small schemas and put them directly into test files so for new rules let's do it from the start.
BTW. Are you interested in working on breaking "harness" schema?

danielrearden replied:
I can do that. Maybe I could knock that out as a separate PR first and then rebase this.

IvanGoncharov replied:
Actually, if you are interested in doing that work, we can merge this first and your later change all the rules in one decisive PR.

danielrearden replied:
That works too!

IvanGoncharov commented:
> Would NoIntrospectionQueriesRule be more inline with the existing rule names?

@danielrearden Generally, yes. But in this case, we forbid something that is totally valid and you ban it because of your own preference/requirements not even because it represents best practice.
Speaking of this can you please expand description with some strong wording asking not to use this rule unless it's absolutely necessary.

danielrearden commented:
@IvanGoncharov 

* Changed name to `ProhibitIntrospectionQueriesRule`
* Updated rule logic to just check the field type
* Moved rule to `/rules/optional`
* Updated comment to include usage warning
* Added test for non-standard introspection fields -- ended up using `expectValidationRulesWithSchema` anyway so I wouldn't have to touch the harness schema 
* Had to build schema programmatically since `buildSchema` doesn't support using introspection types for fields

The use of introspection types with user-defined fields is unconventional but I don't believe it's prohibited by the spec. Since the rule just tests the field's type, it's possible it would also throw in the event a non-standard introspection field like that was requested. I think that behavior lines up with the intent behind the rule.

IvanGoncharov commented on src/validation/rules/optional/ProhibitIntrospectionQueriesRule.js:
```diff
@@ -0,0 +1,47 @@
+          fieldType = fieldType.ofType;
+        }
+
+        if (introspectionTypes.includes(fieldType)) {
```
Why not simply call `isIntrospectionType`?

IvanGoncharov commented on src/validation/rules/optional/ProhibitIntrospectionQueriesRule.js:
```diff
@@ -0,0 +1,47 @@
+        let fieldType = fieldDef.type;
+        while (isWrappingType(fieldType)) {
+          fieldType = fieldType.ofType;
+        }
```
We have `getNamedType` for that

IvanGoncharov commented on src/validation/rules/optional/ProhibitIntrospectionQueriesRule.js:
```diff
@@ -0,0 +1,47 @@
+): ASTVisitor {
+  return {
+    Field(node: FieldNode) {
+      const fieldDef = context.getFieldDef();
```
If you want to just get resulting type please use `getType`

danielrearden commented:
@IvanGoncharov Ok, that should be all of it 😓

IvanGoncharov commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "javascript.validate.enable": false
+}
```
Please remove this file.

danielrearden replied:
Dangit

IvanGoncharov commented on src/validation/__tests__/ProhibitIntrospectionQueriesRule-test.js:
```diff
@@ -0,0 +1,146 @@
+  expectErrors(queryStr).to.deep.equal([]);
+}
+
+describe('Validate: Prohibit introspection queries', () => {
```
Please add one test with introspection field being reported in fragment.

IvanGoncharov commented on src/validation/__tests__/ProhibitIntrospectionQueriesRule-test.js:
```diff
@@ -0,0 +1,146 @@
+    `).to.deep.equal([
+      {
+        locations: [
+          {
```
Please format locations as in the other test: single line with `line` followed by `column`

IvanGoncharov commented on src/validation/__tests__/ProhibitIntrospectionQueriesRule-test.js:
```diff
@@ -0,0 +1,146 @@
+          },
+        ],
+        message:
+          'GraphQL introspection has been disabled, but the requested query contained the field "__schema".',
```
Also, `message` should be first property followed by `locations`

IvanGoncharov commented on src/validation/__tests__/ProhibitIntrospectionQueriesRule-test.js:
```diff
@@ -0,0 +1,146 @@
+      {
+        message:
+          'GraphQL introspection has been disabled, but the requested query contained the field "introspectionField".',
+        locations: [{ line: 4, column: 11 }],
```
Here is good example of correct formatting.

IvanGoncharov commented on src/validation/rules/optional/ProhibitIntrospectionQueriesRule.js:
```diff
@@ -0,0 +1,41 @@
+): ASTVisitor {
+  return {
+    Field(node: FieldNode) {
+      const fieldType = context.getType();
```
You can write it like this:
```js
const type = getNamedType(context.getType());
if (type != null && isIntrospectionType(type) {
```

IvanGoncharov commented on src/validation/rules/optional/ProhibitIntrospectionQueriesRule.js:
```diff
@@ -0,0 +1,41 @@
+import { type ValidationContext } from '../../ValidationContext';
+
+import { getNamedType } from '../../../type/definition';
+import { isIntrospectionType } from '../../../type/introspection';
```
You should order imports in hierarchical so `../../../type` should be before `../../ValidationContext`.
So first ordering criteria.

We put type import value imports in the same import statements so it doesn't make sense to order by them by type/non-type criteria since latter you would mix them in the same statement anyway.

danielrearden replied:
Totally missed that's the code style being enforced. Would it make sense to update the [import/order](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/order.md) ESLint rule to help enforce that?

IvanGoncharov replied:
If you can do that would be great at the same time I don't think we can enforce everything with their current config.
For example, I order imports in order (it matches order of chapters in GraphQL spec):
* polyfills
* \__testUtils__
* jsutils
* error
* **\<same order as chapters in graphql spec>**
* language
* type
* validation
* execution
* subscription
* **\<end of the same order>**
* utilities
* `src/graphql.js`
* files in upper dir
* files in the same dir
* \__fixtures__

If we can enforce something like this with `import/order` I'm totally for it if not I think we need to write our own custom rule.
I'm also ok to change the order if it would be easier to enforce it but for folders that present spec chapters like to keep them in the same order.

@danielrearden Can you please open an issue to track this?

IvanGoncharov replied:
Also, it definitely not compete with TS migration but actually help with reviewing such PRs.

IvanGoncharov commented on src/validation/__tests__/ProhibitIntrospectionQueriesRule-test.js:
```diff
@@ -0,0 +1,146 @@
+          name: 'SomeType',
+          fields: {
+            someField: { type: GraphQLString },
+            introspectionField: { type: __Schema },
```
If you like to test a non-standard introspection field it's better to use some type that is not exposed directly by GraphQL itself, for example, type for enum value.

IvanGoncharov commented on src/validation/__tests__/ProhibitIntrospectionQueriesRule-test.js:
```diff
@@ -0,0 +1,146 @@
+  GraphQLObjectType,
+  GraphQLString,
+  __Schema,
+} from '../../type/index';
```
Same note about order of import, more "external" files should go first.

IvanGoncharov commented on src/validation/index.js:
```diff
@@ -94,3 +94,6 @@ export { UniqueEnumValueNamesRule } from './rules/UniqueEnumValueNamesRule';
 export { PossibleTypeExtensionsRule } from './rules/PossibleTypeExtensionsRule';
+
+// Optional rules not defined by the GraphQL Specification
+export { ProhibitIntrospectionQueriesRule } from './rules/optional/ProhibitIntrospectionQueriesRule';
```
Please duplicate it in `src/validation/index.d.ts`, `src/index.{js,d.ts}`

IvanGoncharov commented:
@danielrearden Forget about this one.
But it too late in my timezone so I will go to 🛌 will check it tomorrow.

P.S. It's totally okay to ping in comments after you addressed all review comments. 
Moreover actually makes my life easier since I don't need to check all current PRs to see their statuses 😄

IvanGoncharov commented on src/validation/__tests__/NoSchemaIntrospectionCustomRule-test.js:
```diff
@@ -0,0 +1,153 @@
+      },
+    },
+  }),
+});
```
@danielrearden I merged #2608 so now `buildSchema` should work fine can you please use it instead.
